alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/Orienting.kt

```diff
@@ -0,0 +1,179 @@
+     * boundary of the current room. This method finds the point of such segment
+     * which is more convenient to cross.
+     */
+    override fun crossingPoint(targetDoor: Euclidean2DPassage): Euclidean2DPosition {
+        with(targetDoor.passageShape) {
+            val nextRoom = environment.graph().getEdgeTarget(targetDoor)
+            /*
+             * The ideal movement the pedestrian would perform connects its current
+             * position to the centroid of the next room.
+             */
+            val desiredMovement = Segment2D(environment.getPosition(pedestrian), nextRoom.centroid)
+            /*
+             * The sub-destination is computed as the point belonging to the door which
+             * is closest to the intersection of the lines defined by the desired
+             * movement and the crossing itself.
+             */
+            val subdestination = closestPointTo(linesIntersection(this, desiredMovement).point.get())
+            /*
+             * If the sub-destination is an end point of the crossing, we move it slightly
+             * away from such end point to avoid behaviors in which the pedestrian always
+             * moves remaining attached to the walls.
+             */
+            if (subdestination == first || subdestination == second) {
+                val correction = if (subdestination == first) {
+                    second - first
+                } else {
+                    first - second
+                }.resized(toVector().magnitude * wallRepulsionFactor)
+                return subdestination + correction
+            }
+            return subdestination
+        }
+    }
+
+    override fun interpolatePositions(
+        current: Euclidean2DPosition,
+        target: Euclidean2DPosition,
+        maxWalk: Double
+    ): Euclidean2DPosition =
+        Seek2D(environment, reaction, pedestrian, *target.coordinates).interpolatePositions(current, target, maxWalk)
+
+    /**
+     * The [congestionFactor] is added.
+     */
+    public override fun weight(edge: Euclidean2DPassage, rank: Int?) =
+        super.weight(edge, rank) * congestionFactor(environment.graph().getEdgeTarget(edge))
+
+    /*
+     * This factor takes into account the congestion of the room the edge being weighted
+     * leads to. It assumes the pedestrian can asses the level of congestion of such room
+     * even if he's not located inside it.
+     */
+    private fun congestionFactor(room: M): Double =
+        environment.nodes
+            .filterIsInstance<Pedestrian<T>>()
+            .filter { room.contains(environment.getPosition(it)) }
+            .count()
+            .let { pedestrian.shape.diameter.pow(2) * it / room.asAwtShape().area() + 1 }
+
+    /*
+     * A rough estimation of the area of a shape.
+     */
+    private fun Shape.area(): Double =
+        with(bounds2D) {
+            abs(width * height)
+        }
+
+    companion object {
+        private const val wallRepulsionFactor = 0.3
```

Comment:
How was this computed?
We need this to be public, documented, and modifiable. Otherwise it's a hidden parameter

Reply:
It was found empirically, I moved it to the constructor and documented it.